src/psyclone/transformations.py

```diff
@@ -1213,6 +1243,81 @@ def _create_colours_loop(self, node):
 
         return colours_loop
 
+    def _create_tiled_colours_loops(self, node: Loop) -> Loop:
+        '''
+        Creates a nested loop hierarchy (colours, tiles and cells of a given
+        colour inside a tile) which can be used to replace the supplied loop
+        over cells.
+
+        If it iterates over halos it will look like:
+
+        do colour = 1, ntilecolours
+          do tile = 1, last_halo_tile_per_colour(colour, hdepth)
+            do cell =  1, last_halo_cell_per_colour_and_tile(tile, colour, hd)
+
+        If it does not iterate over halos it will look like:
+
+        do colour = 1, ntilecolours
+          do tile = 1, last_edge_tile_per_colour(colour)
+            do cell =  1, last_edge_cell_per_colour_and_tile(tile, colour)
+
+        :param node: the loop for which to create a coloured version.
+
+        :returns: doubly-nested loop over colours and cells of a given colour.
+
+        '''
+        # Create a 'colours' loop. This loops over colours and must be run
+        # sequentially.
+        colours_loop = node.__class__(parent=node.parent, loop_type="colours")
+        colours_loop.field_space = node.field_space
+        colours_loop.iteration_space = node.iteration_space
+        colours_loop.set_lower_bound("start")
+        colours_loop.set_upper_bound("ntilecolours")
+
+        # Create a 'colourtile' loop. This loops over tiles of a particular
+        # colour and can be run in parallel.
+        colour_loop = node.__class__(parent=colours_loop.loop_body,
+                                     loop_type="colourtiles")
+        colour_loop.field_space = node.field_space
+        colour_loop.field_name = node.field_name
+        colour_loop.iteration_space = node.iteration_space
+        colour_loop.set_lower_bound("start")
+        colour_loop.kernel = node.kernel
+        if node.upper_bound_name in LFRicConstants().HALO_ACCESS_LOOP_BOUNDS:
+            # If the original loop went into the halo then this coloured loop
+            # must also go into the halo.
+            index = node.upper_bound_halo_depth
+            colour_loop.set_upper_bound("ntiles_per_colour_halo", index)
+        else:
+            # No halo access.
+            colour_loop.set_upper_bound("ntiles_per_colour")
+
+        # Add this loop as a child of our loop over colours
+        colours_loop.loop_body.addchild(colour_loop)
+
+        # Create a cells loop. This loops over cells of a particular tile
+        # and can be run in parallel.
+        tile_loop = node.__class__(parent=colour_loop.loop_body,
+                                   loop_type="tile")
```

Comment:
Possibly "cells_in_tile" would be a clearer loop type than "tile"?

Reply:
It is, it was done in the same style as "colour" but I think "cells_in_colour" is also a clearer name, and "tiles_in_colour" instead of "tilecolour" so I will update the 3 names everywhere.

Reply:
Great, thanks for making that change. I think it's much easier to understand now.